subsys/mgmt/osdp/src/osdp_pd.c

```diff
@@ -635,6 +639,46 @@ static int pd_build_reply(struct osdp_pd *pd, uint8_t *buf, int max_len)
 		buf[len++] = ISSET_FLAG(pd, PD_FLAG_R_TAMPER);
 		ret = OSDP_PD_ERR_NONE;
 		break;
+	case REPLY_KEYPPAD:
+		event = (struct osdp_event *)pd->ephemeral_data;
+		assert_buf_len(REPLY_KEYPAD_LEN + event->keypress.length, max_len);
+		buf[len++] = pd->reply_id;
+		buf[len++] = (uint8_t)event->keypress.reader_no;
+		buf[len++] = (uint8_t)event->keypress.length;
+		for (i = 0; i < event->keypress.length; i++) {
+			buf[len++] = event->keypress.data[i];
+		}
+		ret = OSDP_PD_ERR_NONE;
+		break;
+	case REPLY_RAW: {
+		int len_bytes;
+
+		event = (struct osdp_event *)pd->ephemeral_data;
+		len_bytes = (event->cardread.length + 7) / 8;
+		assert_buf_len(REPLY_RAW_LEN + len_bytes, max_len);
+		buf[len++] = pd->reply_id;
+		buf[len++] = (uint8_t)event->cardread.reader_no;
+		buf[len++] = (uint8_t)event->cardread.format;
+		buf[len++] = BYTE_0(event->cardread.length);
+		buf[len++] = BYTE_1(event->cardread.length);
+		for (i = 0; i < len_bytes; i++) {
+			buf[len++] = event->cardread.data[i];
+		}
```

Comment:
Consider usnig `memcpy` for data copying.

Reply:
Ack, thanks.